src/ApiCompat/Microsoft.DotNet.ApiCompatibility/Rules/AttributesMustMatch.cs

```diff
@@ -0,0 +1,263 @@
+﻿// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+//
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Collections.Specialized;
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+using System.Runtime;
+using System.Text;
+using Microsoft.CodeAnalysis;
+using Microsoft.DotNet.ApiCompatibility.Abstractions;
+
+namespace Microsoft.DotNet.ApiCompatibility.Rules
+{
+    /// <summary>
+    /// This class implements a rule to check that the attributes between public members do not change.
+    /// </summary>
+    public class AttributesMustMatch : IRule
+    {
+        private readonly RuleSettings _settings;
+
+        private class AttributeGroup : IEqualityComparer<AttributeGroup>
+        {
+            public AttributeData? _repr;
+            public List<AttributeData?>? _attributes;
+            public List<bool>? _seen;
+            private RuleSettings _Settings;
+
+            public AttributeGroup(RuleSettings Settings) { _Settings = Settings; }
+            public AttributeGroup(RuleSettings Settings, AttributeData? attr)
+            {
+                _Settings = Settings;
+                _repr = attr;
+                _attributes = new List<AttributeData?>();
+                _seen = new List<bool>();
+                add(attr);
+            }
+
+            public void add(AttributeData? attr)
+            {
+                _attributes?.Add(attr);
+                _seen?.Add(false);
+            }
+
+            public bool Equals(AttributeGroup? x, AttributeGroup? y) => _Settings.SymbolComparer.Equals(x?._repr?.AttributeClass!, y?._repr?.AttributeClass!);
+            public int GetHashCode(AttributeGroup? obj) => _Settings.SymbolComparer.GetHashCode(obj?._repr?.AttributeClass!);
+        }
+        private class AttributeSet : IEnumerable<AttributeGroup>
+        {
+            private HashSet<AttributeGroup> _set;
+            private readonly RuleSettings _settings;
+
+            public AttributeSet(RuleSettings Settings, IList<AttributeData> attributes)
+            {
+                _set = new HashSet<AttributeGroup>(new AttributeGroup(Settings));
+                _settings = Settings;
+                for (int i = 0; i < attributes.Count; i++)
+                {
+                    add(attributes[i]);
+                }
+            }
+
+            public void add(AttributeData attr)
```

Comment:
Public methods should start with an upper case letter in C#. Same for the other methods below.